bin/vernemq.sh

```diff
@@ -61,19 +61,21 @@ if env | grep "DOCKER_VERNEMQ_DISCOVERY_KUBERNETES" -q; then
     sed -i.bak -r "s/VerneMQ@.+/VerneMQ@${VERNEMQ_KUBERNETES_HOSTNAME}/" /vernemq/etc/vm.args
     # Hack into K8S DNS resolution (temporarily)
     kube_pod_names=$(curl -X GET $insecure --cacert /var/run/secrets/kubernetes.io/serviceaccount/ca.crt https://kubernetes.default.svc.$DOCKER_VERNEMQ_KUBERNETES_CLUSTER_NAME/api/v1/namespaces/$DOCKER_VERNEMQ_KUBERNETES_NAMESPACE/pods?labelSelector=$DOCKER_VERNEMQ_KUBERNETES_LABEL_SELECTOR -H "Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)" | jq '.items[].spec.hostname' | sed 's/"//g' | tr '\n' ' ')
-    for kube_pod_name in $kube_pod_names;
-    do
-        if [ $kube_pod_name == "null" ]
-            then
-                echo "Kubernetes discovery selected, but no pods found. Maybe we're the first?"
-                echo "Anyway, we won't attempt to join any cluster."
-                break
+    for kube_pod_name in $kube_pod_names; do
+        if [ $kube_pod_name == "null" ]; then
+            echo "Kubernetes discovery selected, but no pods found. Maybe we're the first?"
+            echo "Anyway, we won't attempt to join any cluster."
+            break
         fi
-        if [ $kube_pod_name != $MY_POD_NAME ]
-            then
-                echo "Will join an existing Kubernetes cluster with discovery node at ${kube_pod_name}.${VERNEMQ_KUBERNETES_SUBDOMAIN}.${DOCKER_VERNEMQ_KUBERNETES_NAMESPACE}.svc.${DOCKER_VERNEMQ_KUBERNETES_CLUSTER_NAME}"
-                echo "-eval \"vmq_server_cmd:node_join('VerneMQ@${kube_pod_name}.${VERNEMQ_KUBERNETES_SUBDOMAIN}.${DOCKER_VERNEMQ_KUBERNETES_NAMESPACE}.svc.${DOCKER_VERNEMQ_KUBERNETES_CLUSTER_NAME}')\"" >> /vernemq/etc/vm.args
-                break
+        if [ $kube_pod_name != $MY_POD_NAME ]; then
+            echo "Will join an existing Kubernetes cluster with discovery node at ${kube_pod_name}.${VERNEMQ_KUBERNETES_SUBDOMAIN}.${DOCKER_VERNEMQ_KUBERNETES_NAMESPACE}.svc.${DOCKER_VERNEMQ_KUBERNETES_CLUSTER_NAME}"
+            echo "-eval \"vmq_server_cmd:node_join('VerneMQ@${kube_pod_name}.${VERNEMQ_KUBERNETES_SUBDOMAIN}.${DOCKER_VERNEMQ_KUBERNETES_NAMESPACE}.svc.${DOCKER_VERNEMQ_KUBERNETES_CLUSTER_NAME}')\"" >>/vernemq/etc/vm.args
```

Comment:
Today during a redeployment of an already existing cluster we faced the following corner case which resulted in 2 different VerneMQ Clusters.
1st cluster
```
vernemq-0
```
2nd cluster
```
vernemq-1
vernemq-2
```

Due to `if [ $kube_pod_name != $MY_POD_NAME ]; then` the following will happen
```
vernemq-0 will try to connect to vernemq-1
vernemq-1 will try to connect to vernemq-0
vernemq-2 will try to connect to vernemq-0
```
So during the redeployment `vernemq-2` is firstly restarted and tried to connect to the node mentioned above, which is successful and continues as it should.
To continue with `vernemq-1` is then restarted and it tries to connect to the node mentioned above but faced the below
```
15:46:42.860 [info] Sent join request to: 'VerneMQ@vernemq-0.vernemq-headless.vernemq.svc.cluster.local'
15:46:42.863 [info] Unable to connect to 'VerneMQ@vernemq-0.vernemq-headless.vernemq.svc.cluster.local'
```
for some strange reason the `vernemq-0` pod has already started the update as well during this time period and faced the same error
```
15:46:50.307 [info] Sent join request to: 'VerneMQ@vernemq-1.vernemq-headless.vernemq.svc.cluster.local'
15:46:50.309 [info] Unable to connect to 'VerneMQ@vernemq-1.vernemq-headless.vernemq.svc.cluster.local'
```
For some strange reason we ended up with 2 clusters.
As soon as we deleted the `vernemq-0` it connected correctly to the correct VerneMQ cluster and joined the other nodes.
Unfortunately we didn't keep the logs of that due to being in a hurry to fix the issue.
Any thoughts?

Reply:
@angeloskaltsikis you're calling this re-deployment but also upgrade. What exactly did you try to do? 
Is there some sort of internal name caching or dns caching in K8s going on here maybe?